Introduce a script for checking file permissions

shindere opened the pull request:
this GPR is a follow-up to the discussion that took place in GPR#2270.

The proposed script makes sure files with an executable permission really
diserve it, meaning they are either boot/ocaml{c,lex} or scripts
starting with #!.

It will then become possible to use this script in our various CI
solutions but this can happen in future PRs.

Any comment regarding the implementation is welcome.
In particular, the script only looks for permissions like 755,
perhaps this should be improved e.g. with a regexp?

XVilka commented:
Maybe it makes sense to set it as a git pre-commit hook as well?

shindere commented:
Anton Kochkov (2019/03/01 08:22 +0000):
> Maybe it make sense to set it as a git hook as well?

Perhaps, I don't know. Would it be okay, in a hook, to test the whole
repo rather than just the files that are modified by the commit? Because
you mean a commit hook, right?

dra27 commented:
@shindere - thanks for writing the script! It would be good to update our pre-commit githook too, which already has the necessary machinery just to examine the files in the commit - I’ll push something later, if you don’t get there first!

shindere commented:
David Allsopp (2019/03/01 00:56 -0800):
> @shindere - thanks for writing the script! It would be good to update
> our pre-commit githook too, which already has the necessary machinery
> just to examine the files in the commit - I’ll push something later,
> if you don’t get there first!

I'd prefer if you can have a look at it, thanks!

It could even be done in another PR.

shindere commented:
Rebased and updated to take into account a lunch-time comment given by
@xavierleroy:

It is actually important that
`testsuite/tests/lib-unix/unix-execvpe/subdir/script2`
does not begin with a #! because the very purppose of the test it
belongs to is to check how exec behaves on such files.

xavierleroy commented:
Thank you for the attention you gave to `testsuite/tests/lib-unix/unix-execvpe/subdir/script2`.  The `exec` family of system calls behave slightly differently on scripts that start with `#!/bin/sh` than on scripts that don't start with `#!`, so it is intentional to test both.

shindere commented:
Okay, many thanks for having clarified this!

xavierleroy reviewed (commented):
Looks good to me.  I'm sure someone else will suggest the use of git attributes to mark the exceptions, but I don't mind.

I'm the one who suggested the name `check-file-modes` for this script but perhaps `check-file-permissions` would be more idiomatic.

shindere commented:
Xavier Leroy (2019/03/01 05:46 -0800):
> xavierleroy commented on this pull request.
> 
> Looks good to me.  I'm sure someone else will suggest the use of git
> attributes to mark the exceptions, but I don't mind.

We discussed it with @gasche yesterday and agreed it's okay to go ahead
that way for now, we can always do it differently later if the need
appears.

> I'm the one who suggested the name `check-file-modes` for this script
> but perhaps `check-file-permissions` would be more idiomatic.

Done. Didn't rename the branch though.

xavierleroy commented on tools/check-file-permissions:
```diff
@@ -0,0 +1,46 @@
+
+check_script()
+{
+  if [ x`head -c2 $1` != x'#!' ]; then
```
Just to ruin your day: the `-c` option to `head` is not standard (see https://pubs.opengroup.org/onlinepubs/9699919799/utilities/head.html#tag_20_57) and is not implemented in OpenBSD.  (But FreeBSD and MacOS have it.)  So, your script will not run everywhere.  

Possible alternative: `if head -n 1 "$1" | grep -q '^#!' `.

xavierleroy commented on tools/check-file-permissions:
```diff
@@ -0,0 +1,46 @@
+{
+  if [ x`head -c2 $1` != x'#!' ]; then
+    ok=false;
+    echo $1 does not seem to diserve its executable permission
```
"deserve", not "diserve".

xavierleroy reviewed (changes requested):
Typo + small portability issue.

shindere commented:
Well, thanks for the useful feedback! I have to say, it didn't even come
to me that there could be a portability issue with this option of
`head` so many thanks for having thought about it.


Proposed fixes applied, thanks again!

dra27 commented:
Just to say that I've got an update to `check-typo` to do this, as that propagates to the githook … just debugging one weirdness with the awk script (bashes head against desk)

xavierleroy commented on tools/check-file-permissions:
```diff
@@ -0,0 +1,46 @@
+
+check_script()
+{
+  if head -n 1 "$1" | grep -q '^#!'; then
```
You need to invert the test.  It's not because I suggest something that you should not think before implementing the suggestion :-)

shindere committed:
Introduce a script for checking file permissions

The tools/check-file-permissions script reports files that have
an executable permission but do not seem to deserve it, i.e. they are
neither a shell script nor one of these exceptions:
boot/ocaml{c,lex}, testsuite/tests/lib-unix/unix-execvpe/subdir/script2.

Note that the later file is a script but for the very purpose of
the test it belongs to, it is fundamental that it does not start with #!.

shindere commented:
Xavier Leroy (2019/03/02 02:37 -0800):
> You need to invert the test.  It's not because I suggest something
> that you should not think before implementing the suggestion :-)

Blind devotion and its dangers...

It's fixed and tested, this time, thanks, and sorry.

shindere commented:
This is superseeded by #2298, thus closing.
